### DISCUSSION TITLE
AlwaysMaybeNull attribute for concurrently modified members.

### DISCUSSION THREAD

**Discussion by Kuinox** [dotnet/csharplang#6015] [Language Ideas]
Nullable attributes currently give a false sense of security in concurrent code, example:

```csharp
object? _field; //may be concurrently modified elsewhere.

public string DoBugged()
{
    if(_field is not null)
    {
         return _field.ToString(); // Potential nullref exception.
    }
}

public string DoCorrect()
{
   var field = _field;
    if(field is not null)
    {
         return field.ToString();
    }
}
```

A new attribute, AlwaysMaybeNull, would make the marked field/property "maybe null" even after a nullability check.
The correct way to use this field would be either by copying the reference then use it (like DoCorrect does), or using safe operators like `?.` (which do exactly this).

There is further enhancement possible to be made on top of it, like better interaction with `ref` and `in` arguments.

**Comment by RikkiGibson** [comment]
IMO, this case is best suited to "concurrency safety" analyzer of some kind, not something in the language.

**Reply by 333fred** [reply]
Agreed. For example, you might have a struct that needs to be copied to the stack before reading to avoid seeing sheared values.

**Reply by Kuinox** [reply]
Indeed a "Must Copy" attributes with his own analyzers would also cover this use case.
